frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadButton.tsx

```diff
@@ -1,29 +1,62 @@
-import { useCallback } from 'react'
-import { useMutation } from 'react-query'
-import { Box, MenuButton } from '@chakra-ui/react'
+import { useCallback, useMemo, useState } from 'react'
+import { useThrottle } from 'react-use'
+import { Box, MenuButton, Text, useDisclosure } from '@chakra-ui/react'
 
 import { BxsChevronDown } from '~assets/icons/BxsChevronDown'
 import { BxsChevronUp } from '~assets/icons/BxsChevronUp'
+import { useTimeout } from '~hooks/useTimeout'
 import Badge from '~components/Badge'
 import Button from '~components/Button'
 import Menu from '~components/Menu'
 
 import { useStorageResponsesContext } from '../StorageResponsesContext'
-import useDecryptionWorkers, {
-  DownloadEncryptedParams,
-} from '../useDecryptionWorkers'
+import useDecryptionWorkers from '../useDecryptionWorkers'
+
+import { DownloadWithAttachmentModal } from './DownloadWithAttachmentModal'
+import { ProgressModal } from './ProgressModal'
 
 export const DownloadButton = (): JSX.Element => {
-  const { downloadEncryptedResponses } = useDecryptionWorkers()
-  const { downloadParams } = useStorageResponsesContext()
+  const {
+    isOpen: isDownloadModalOpen,
+    onClose: onDownloadModalClose,
+    onOpen: onDownloadModalOpen,
+  } = useDisclosure()
+  const {
+    isOpen: isProgressModalOpen,
+    onClose: onProgressModalClose,
+    onOpen: onProgressModalOpen,
+  } = useDisclosure()
 
-  const handleExportCsvMutation = useMutation(
-    (params: DownloadEncryptedParams) => downloadEncryptedResponses(params),
-    // TODO: add error and success handling
-  )
+  const [progressModalTimeout, setProgressModalTimeout] = useState<
+    number | null
+  >(null)
+  const { downloadParams, responsesCount } = useStorageResponsesContext()
+  const [_downloadCount, setDownloadCount] = useState(0)
+  const downloadCount = useThrottle(_downloadCount, 1000)
+
+  const downloadPercentage = useMemo(() => {
+    if (!responsesCount) return 0
+    return Math.floor((downloadCount / responsesCount) * 100)
+  }, [downloadCount, responsesCount])
+
+  useTimeout(onProgressModalOpen, progressModalTimeout)
```

Comment:
so the timeout is initially `null` which doesn't do anything, and then the timeout gets set when `setProgressModalTimeout` is called and hence `progressModalTimeout` changes?

Reply:
yes